### ISSUE TITLE
[help] I want to write a helm source. In the action, I want to update candidates but don't quit helm. Is there documents for this?

### ISSUE THREAD

**Issue by netjune** [emacs-helm/helm#2401]
I can't find documents for this.

**Comment by thierryvolpiatto** [comment]
zhanghj ***@***.***> writes:

> I can't find documents for this.

I don't know exactly what you want to do, but it seems you just want to
update your candidates without quitting helm, for this you don't have to
create an action but use C-c C-u (force-update).
Here an example:

    (defvar helm-update-example-cache nil)

    (defun helm-update-example-get-candidate ()
      (or helm-update-example-cache
          (setq helm-update-example-cache
                (cl-loop repeat 10
                         collect (random 10)))))

    (helm :sources (helm-build-sync-source "test"
                     :candidates 'helm-update-example-get-candidate
                     :update (lambda () (setq helm-update-example-cache nil)))
          :buffer "*helm test*")

Eval the last block and see that at each time you hit C-c C-u your
candidates are refreshed.

See documentation of :update slot in helm-source docstring.

NOTE: To have a readable view of helm-source docstring don't use regular
emacs describe-function but helm-apropos, Emacs formatting of class
definitions is terrible and unreadable.

Hope that helps.

> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.*


-- 
Thierry

**Comment by netjune** [comment]
`I don't know exactly what you want to do, but it seems you just want to
update your candidates without quitting helm, for this you don't have to
create an action but use C-c C-u (force-update).`

Sorry, I didn't discribe the question clearly.

I want to implement a command based on helm. It is like helm-for-files, but has some features. The main two features:

1. The candidates only includes directories and certain files types(orgmode file and markdown file). Other files are ignored.
2. I can enter a directory in candidates or go to parent directory(..) without quitting helm. I think this needs to update candidates in the action.

Thanks.

**Comment by thierryvolpiatto** [comment]
zhanghj ***@***.***> writes:

> I don't know exactly what you want to do, but it seems you just want to update your candidates without quitting helm, for this you don't have to create an action but use C-c C-u
> (force-update).
>
> Sorry, I didn't discribe the question clearly.
>
> I want to implement a command based on helm. It is like helm-for-files, but has some features. The main two features:
>
>  1. The candidates only includes directories and certain files types(orgmode file and markdown file). Other files are ignored.
>  2. I can enter a directory in candidates or go to parent directory(..) without quitting helm. I think this needs to update candidates in the action.
>
> Thanks.

This is already implemented e.g.

    (read-file-name "test: "
                    nil nil nil nil
                    (lambda (f)
                      (or (file-directory-p f)
                          (helm-aand (file-name-extension f)
                                     (string-match "\\(org\\|md\\)\\'" it)))))

Don't forget to enable helm-mode.

The implementation is more or less helm-find-files (see
helm-source-ffiles defclass in helm-files.el),
basically the candidate function is following helm-pattern and call
directory-files on pattern (or its base directory).
You should use a sync source and volatile slot to implement your source,
but why reinventing the wheel when it's already implemented? ;-)


-- 
Thierry

**Comment by netjune** [comment]
Yes, it works. I see helm-find-files. It is complex. I want to write a simple version. I known how to prepare candidates. But I don't konwn how to reuse helm session.

**Comment by thierryvolpiatto** [comment]
zhanghj ***@***.***> writes:

> Yes, it works. I see helm-find-files. It is complex. I want to write a
> simple version.

Here a simple example that may be a starting point:

    (helm :sources (helm-build-sync-source "test"
                     :candidates (lambda ()
                                   (let ((dir (if (file-directory-p helm-pattern)
                                                  helm-pattern
                                                (file-name-directory helm-pattern))))
                                     (directory-files dir t)))
                     :volatile t)
          :buffer "*helm test*"
          :input default-directory)

> I known how to prepare candidates. But I don't konwn how to reuse helm
> session.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.*


-- 
Thierry

**Comment by netjune** [comment]
> 
> zhanghj ***@***.***> writes:
> Yes, it works. I see helm-find-files. It is complex. I want to write a simple version.
> Here a simple example that may be a starting point: (helm :sources (helm-build-sync-source "test" :candidates (lambda () (let ((dir (if (file-directory-p helm-pattern) helm-pattern (file-name-directory helm-pattern)))) (directory-files dir t))) :volatile t) :buffer "*helm test*" :input default-directory)
> […](#)
> I known how to prepare candidates. But I don't konwn how to reuse helm session. — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.*
> -- Thierry

Thanks. When I select a directory and press ENTER, helm quits. I want to keep the helm session, and just update candidates(files in the selected directory). Is it possible? Do I need to restart helm in the action?

**Comment by thierryvolpiatto** [comment]
zhanghj ***@***.***> writes:

>     zhanghj @.**> writes:
>     Yes, it works. I see helm-find-files. It is complex. I want to write a simple version.
>     Here a simple example that may be a starting point: (helm :sources (helm-build-sync-source "test" :candidates (lambda () (let ((dir (if (file-directory-p helm-pattern) helm-pattern
>     (file-name-directory helm-pattern)))) (directory-files dir t))) :volatile t) :buffer "helm test" :input default-directory)
>     …
>     I known how to prepare candidates. But I don't konwn how to reuse helm session. — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or
>     unsubscribe.
>     -- Thierry
>    
> Thanks. When I select a directory and press ENTER, helm quits.

Of course it is just an example with no action (action is 'identity by default).

> I want to keep the helm session, and just update candidates(files in
> the selected directory). Is it possible? Do I need to restart helm in
> the action?

Please read https://github.com/emacs-helm/helm/wiki/Developing

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.*


-- 
Thierry

**Comment by netjune** [comment]
> Please read https://github.com/emacs-helm/helm/wiki/Developing

I have looked at it. But I didn't find the answer.

I looked into `helm-source-ffiles` and its persistant-action-if `helm-find-files-persistent-action-if`, in which the work(pressing ENTER to expand the directory) is done.

I see in document that the persistant action is triggered by C-j/C-z. Why it is triggered by ENTER in helm-find-files?

**Comment by thierryvolpiatto** [comment]
zhanghj ***@***.***> writes:

>     Please read https://github.com/emacs-helm/helm/wiki/Developing
>    
> I have looked at it. But I didn't find the answer.

You didn't look carefully, see
https://github.com/emacs-helm/helm/wiki/Developing#writing-persistent-actions

-- 
Thierry

**Comment by netjune** [comment]
> 
> zhanghj ***@***.***> writes:
> Please read https://github.com/emacs-helm/helm/wiki/Developing I have looked at it. But I didn't find the answer.
> You didn't look carefully, see https://github.com/emacs-helm/helm/wiki/Developing#writing-persistent-actions
> […](#)
> -- Thierry

I look again and can't find the anwser. Then I find that in `helm-find-files`, ENTER(RET) is bound to helm-ff-RET which calls `helm-execute-persistent-action` which may update candidates and reuse helm session. Is it?

**Comment by thierryvolpiatto** [comment]
zhanghj ***@***.***> writes:

>     zhanghj @.***> writes:
>     Please read https://github.com/emacs-helm/helm/wiki/Developing I have looked at it. But I didn't find the answer.
>     You didn't look carefully, see https://github.com/emacs-helm/helm/wiki/Developing#writing-persistent-actions
>     …
>     -- Thierry
>    
> I look again and can't find the anwser. Then I find that in helm-find-files, ENTER(RET) is bound to helm-ff-RET which calls helm-execute-persistent-action which may update candidates and
> reuse helm session. Is it?

Yes.

>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.*


-- 
Thierry

**Comment by twlz0ne** [comment]
@thierryvolpiatto 

I wrote some code to simulate the helm ff and it seems to work, but I didn't use the keyword `:persistent-action` and `:persistent-action-if`, is there something wrong?

``` elisp
(defun test-get-cands ()
  (if (string-prefix-p "dir2 >" helm-input)
      (list "file21" "file22")
    (list "file1" "dir2 >" "file3")))

(defun test-persistent-action (cand)
  (with-selected-window (or (active-minibuffer-window) (minibuffer-window))
    (unless (or (helm-follow-mode-p) helm--temp-follow-flag)
      (delete-minibuffer-contents)
      (set-text-properties 0 (length cand) nil cand)
      (insert cand)))
  (helm-check-minibuffer-input))

(defun test-action-RET ()
  (interactive)
  (with-helm-alive-p
    (let ((cand (helm-get-selection)))
      (if (equal cand "dir2 >")
          (progn
            (helm-attrset 'test-action '(test-persistent-action . never-split))
            (helm-execute-persistent-action 'test-action))
        (prog1 (print cand)
          (helm-exit-minibuffer))))))

(defun test-helm ()
  (interactive)
  (helm :sources
        (helm-build-sync-source "Test"
          :candidates 'test-get-cands
          :match (lambda (cand)
                   (let ((patt (s-chop-prefix "dir2 >" helm-pattern)))
                     (when (string-match-p patt cand)
                       cand)))
          :volatile t
          :keymap (let ((map (make-sparse-keymap)))
                    (set-keymap-parent map helm-map)
                    (define-key map (kbd "RET") 'test-action-RET)
                    map))))
```

![Screenshot_2021-04-25_at_10 00 41_AM-shadow_+3+3-0-0](https://user-images.githubusercontent.com/2653486/115978439-f8fc2280-a5b1-11eb-954a-a78afdd0837f.png)

![Screenshot_2021-04-25_at_10 01 09_AM-shadow_+3+3-0-0](https://user-images.githubusercontent.com/2653486/115978441-fa2d4f80-a5b1-11eb-861c-28dc5dd8e59e.png)

**Comment by thierryvolpiatto** [comment]
twlz0ne ***@***.***> writes:

> @thierryvolpiatto
>
> I wrote some code to simulate the helm ff and it seems to work, but I
> didn't use the keyword :persistent-action and :persistent-action-if,
> is there something wrong?

No, passing the attribute name to helm-execute-persistent-action is fine.

-- 
Thierry

**Comment by netjune** [comment]
> 
> 
> @thierryvolpiatto
> 
> I wrote some code to simulate the helm ff and it seems to work, but I didn't use the keyword `:persistent-action` and `:persistent-action-if`, is there something wrong?
> 
> ```emacs-lisp
> (defun test-get-cands ()
>   (if (string-prefix-p "dir2 >" helm-input)
>       (list "file21" "file22")
>     (list "file1" "dir2 >" "file3")))
> 
> (defun test-persistent-action (cand)
>   (with-selected-window (or (active-minibuffer-window) (minibuffer-window))
>     (unless (or (helm-follow-mode-p) helm--temp-follow-flag)
>       (delete-minibuffer-contents)
>       (set-text-properties 0 (length cand) nil cand)
>       (insert cand)))
>   (helm-check-minibuffer-input))
> 
> (defun test-action-RET ()
>   (interactive)
>   (with-helm-alive-p
>     (let ((cand (helm-get-selection)))
>       (if (equal cand "dir2 >")
>           (progn
>             (helm-attrset 'test-action '(test-persistent-action . never-split))
>             (helm-execute-persistent-action 'test-action))
>         (prog1 (print cand)
>           (helm-exit-minibuffer))))))
> 
> (defun test-helm ()
>   (interactive)
>   (helm :sources
>         (helm-build-sync-source "Test"
>           :candidates 'test-get-cands
>           :match (lambda (cand)
>                    (let ((patt (replace-regexp-in-string "^dir2 >" "" helm-pattern)))
>                      (when (string-match-p patt cand)
>                        cand)))
>           :volatile t
>           :keymap (let ((map (make-sparse-keymap)))
>                     (set-keymap-parent map helm-map)
>                     (define-key map (kbd "RET") 'test-action-RET)
>                     map))))
> ```
> 
> ![Screenshot_2021-04-25_at_10 00 41_AM-shadow_+3+3-0-0](https://user-images.githubusercontent.com/2653486/115978439-f8fc2280-a5b1-11eb-954a-a78afdd0837f.png)
> 
> ![Screenshot_2021-04-25_at_10 01 09_AM-shadow_+3+3-0-0](https://user-images.githubusercontent.com/2653486/115978441-fa2d4f80-a5b1-11eb-861c-28dc5dd8e59e.png)

You did what I want to do. :)

**Comment by netjune** [comment]
> 
> twlz0ne ***@***.***> writes:
> @thierryvolpiatto I wrote some code to simulate the helm ff and it seems to work, but I didn't use the keyword :persistent-action and :persistent-action-if, is there something wrong?
> No, passing the attribute name to helm-execute-persistent-action is fine.
> […](#)
> -- Thierry

How to define a local variable for a helm session? It only exists in the helm session. It can be accessed in actions of every source.

**Comment by thierryvolpiatto** [comment]
See helm-set-local-variable.
